### PR TITLE
Add security tracker links to references

### PR DESCRIPTION
**Pull request by hogo6002**
Includes Alpine and Debian security tracker links to vulnerability references on OSV.dev
Fixes https://github.com/google/osv.dev/issues/2284

### CODE REVIEW

**Review by andrewpollock — APPROVED** [review]
LGTM

Are the related links guaranteed to exist, or is it worthwhile to make a HTTP HEAD request to validate it prior to adding it, to avoid adding deadlinks?

File: vulnfeeds/cmd/combine-to-osv/main.go [hunk 1]
```diff
@@ -241,3 +234,41 @@ func loadAllCVEs(cvePath string) map[cves.CVEID]cves.Vulnerability {
...
+
+		// Send the request
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			Logger.Warnf("Unable to validate the URL '%s'", securityReference.URL)
+			return
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode == http.StatusOK {
```
Comment (andrewpollock) [inline vulnfeeds/cmd/combine-to-osv/main.go:267]: See https://github.com/google/osv.dev/blob/260a49adcac4d7d06d75b2f53c0915225e15518e/vulnfeeds/cves/versions.go#L606 for how to do retries if you want to get all fancy and be fault tolerant of intermittent 500 errors...

Comment (hogo6002) [reply vulnfeeds/cmd/combine-to-osv/main.go]: Thanks! Updated.

Comment (andrewpollock) [reply vulnfeeds/cmd/combine-to-osv/main.go]: Give you're already importing cves, can you avoid embedding the code from `cves.ValidateAndCanonicalizeLink()` and use it directly? I think the function is generic enough to meet your needs here.

Comment (hogo6002) [reply vulnfeeds/cmd/combine-to-osv/main.go]: The only issue with `cves.ValidateAndCanonicalizeLink()` is that it internally makes an `http.Head()` request, which fails (with a `406` response code) for Alpine Linux URLs like `https://security.alpinelinux.org/vuln/CVE-2024-32760`. These URLs require an `(Accept: text/html)` header to be sent. I will need to modify the existing validation method to use `http.NewRequest()` and include the necessary header.

```
curl -I https://security.alpinelinux.org/vuln/CVE-2024-32760
HTTP/2 406 
content-type: text/html; charset=utf-8

curl -I --header "Accept: */*" https://security.alpinelinux.org/vuln/CVE-2024-32760
HTTP/2 406 
content-type: text/html; charset=utf-8

curl -I --header "Accept: text/html" https://security.alpinelinux.org/vuln/CVE-2024-32760
HTTP/2 200 
content-type: text/html; charset=utf-8
```

File: vulnfeeds/cves/versions.go [hunk 2]
```diff
@@ -611,7 +611,16 @@ func ValidateAndCanonicalizeLink(link string) (canonicalLink string, err error)
 	}
 	backoff := retry.NewExponential(1 * time.Second)
 	if err := retry.Do(context.Background(), retry.WithMaxRetries(3, backoff), func(ctx context.Context) error {
-		resp, err := http.Head(link)
+		req, err := http.NewRequest("HEAD", link, nil)
+		if err != nil {
+			return err
+		}
+
+		// security.alpinelinux.org responds with text/html content.
```
Comment (another-rex) [inline vulnfeeds/cves/versions.go:619]:
```suggestion
		// security.alpinelinux.org responds with text/html content.
		// default HEAD request in Go does not provide any Accept headers, causing a 406 response
```

Added some clarification for myself.

File: vulnfeeds/cmd/combine-to-osv/main.go [hunk 3]
```diff
@@ -221,3 +233,25 @@ func loadAllCVEs(cvePath string) map[cves.CVEID]cves.Vulnerability {
 	}
 	return result
 }
+
+// addReference adds the related security tracker URL to a given vulnerability's references
+func addReference(cveId string, ecosystem string, convertedCve *vulns.Vulnerability) {
+	var securityReference vulns.Reference
+	securityReference.Type = "ADVISORY"
```
Comment (another-rex) [inline vulnfeeds/cmd/combine-to-osv/main.go:240]: nit, directly initialise securityReference with its Type.

E.g. securityReference := vulns.Reference {
Type: "ADVISORY"
}

File: vulnfeeds/cmd/combine-to-osv/main.go [hunk 4]
```diff
@@ -159,9 +160,20 @@ func combineIntoOSV(loadedCves map[cves.CVEID]cves.Vulnerability, allParts map[c
 				convertedCve.Withdrawn = modified
 			}
 		}
+
+		addedDebianURL := false
+		addedAlpineURL := false
 		for _, pkgInfo := range allParts[cveId] {
 			convertedCve.AddPkgInfo(pkgInfo)
+			if strings.HasPrefix(pkgInfo.Ecosystem, "Debian") && !addedDebianURL {
```
Comment (another-rex) [inline vulnfeeds/cmd/combine-to-osv/main.go:168]: Set these "Debian" and "Alpine" strings to be constants. This way we can get type checking in the if statements in addReference as well.

File: vulnfeeds/cmd/combine-to-osv/main.go [hunk 5]
```diff
@@ -221,3 +233,25 @@ func loadAllCVEs(cvePath string) map[cves.CVEID]cves.Vulnerability {
 	}
 	return result
 }
+
+// addReference adds the related security tracker URL to a given vulnerability's references
+func addReference(cveId string, ecosystem string, convertedCve *vulns.Vulnerability) {
+	var securityReference vulns.Reference
+	securityReference.Type = "ADVISORY"
+	if ecosystem == "Alpine" {
+		securityReference.URL = fmt.Sprintf("https://security.alpinelinux.org/vuln/%s", cveId)
```
Comment (another-rex) [inline vulnfeeds/cmd/combine-to-osv/main.go:242]: Set these URLs as constants.

**Review by another-rex — APPROVED** [review]
Some minor comments, otherwise LGTM

File: vulnfeeds/cmd/combine-to-osv/main.go (hunk 1 above)
Comment (andrewpollock) [reply vulnfeeds/cmd/combine-to-osv/main.go]: That seems safe and backwards compatible to do (and unfortunately annoying)

**Review by andrewpollock — APPROVED** [review]
LGTM and +1 to Rex's feedback

**Review by oliverchang — APPROVED** [review]

**Pull request merged**